### PR TITLE
Issue298 Validator false positive for component imports

### PR DESCRIPTION
**Pull request by kerimoyle**
Addresses bug discussed in #298 

False positives returned from the validateModel when checking components.
The bug is fixed and the test is updated.

### CODE REVIEW

**Comment by kerimoyle** [comment]
@hsorby @agarny How can I re-trigger the coverage checks now that they're working again?

**Comment by hsorby** [comment]
The only way you can do it is to make a change to the code and add it to the pull request.

Some lucky others can re-trigger the CI if they know where to look.

**Comment by kerimoyle** [comment]
This one is ready for review :)

**Comment by agarny** [comment]
> Some lucky others can re-trigger the CI if they know where to look.

FWIW, the way to do it is:
- Check the commit id of what used to be your last commit, i.e. 5c6a0bc;
- Click on the project's [Settings tab](https://github.com/cellml/libcellml/settings), then on [Webhooks](https://github.com/cellml/libcellml/settings/hooks) and finally on the [link](https://github.com/cellml/libcellml/settings/hooks/4594930) for `https://autotest.bioeng.auckland.ac.nz/buildbot/libcellml/change_hook/github/` (assuming you have access to that page);
- Go through the list of `Recent Deliveries` and look for the one that mentions your commit id; and
- Click on the `Redeliver` button.

Voila! :)

P.S.: your commit 5c6a0bc was originally delivered 5 days ago, so you could indeed look only for the deliveries that occurred on 14 June, which would show that the delivery you were after was `3ef3ce60-8e89-11e9-86fb-0a883b58fbe2`.

**Comment by nickerso** [comment]
> > Some lucky others can re-trigger the CI if they know where to look.

> FWIW, the way to do it is:

but note that only repo admins have access to this.

File: tests/validator/validator.cpp [hunk 1]
```diff
@@ -334,32 +334,71 @@ TEST(Validator, importComponents)
...
-    importedComponent3->setSourceComponent(imp3, "component_in_that_model");
+    importedComponent3->setName("invalid_imported_component_in_this_model");
+    importedComponent3->setSourceComponent(imp3, "");
     m->addComponent(importedComponent3);
     v.validateModel(m);
-    EXPECT_EQ(size_t(4), v.errorCount());
+    EXPECT_EQ(size_t(3), v.errorCount());
 
-    // Invalid component import - unnamed component
+    // Invalid component import - duplicate refs
```
Comment (nickerso) [inline tests/validator/validator.cpp]: This is valid, right? the error count doesn't increase...

File: tests/validator/validator.cpp [hunk 2]
```diff
@@ -334,32 +334,71 @@ TEST(Validator, importComponents)
...
-    imp4->setUrl("some-other-different-model.xml");
+    imp4->setUrl("some-other-model.xml");
     libcellml::ComponentPtr importedComponent4 = std::make_shared<libcellml::Component>();
+    importedComponent4->setName("duplicate_imported_component_in_this_model");
     importedComponent4->setSourceComponent(imp4, "component_in_that_model");
     m->addComponent(importedComponent4);
     v.validateModel(m);
+    EXPECT_EQ(size_t(3), v.errorCount());
+
+    libcellml::ImportSourcePtr imp5 = std::make_shared<libcellml::ImportSource>();
```
Comment (nickerso) [inline tests/validator/validator.cpp:367]: comment missing to define this as an invalid import due to missing local name for the imported component.

File: tests/validator/validator.cpp [hunk 3]
```diff
@@ -334,32 +334,71 @@ TEST(Validator, importComponents)
...
+    imp5->setUrl("some-other-different-model.xml");
+    libcellml::ComponentPtr importedComponent5 = std::make_shared<libcellml::Component>();
+    importedComponent5->setSourceComponent(imp5, "component_in_that_model");
+    m->addComponent(importedComponent5);
+    v.validateModel(m);
+    EXPECT_EQ(size_t(5), v.errorCount());
+
+    // Invalid: duplicating component_ref and source
```
Comment (nickerso) [inline tests/validator/validator.cpp]: This one is also valid, right?

File: tests/validator/validator.cpp [hunk 4]
```diff
@@ -334,32 +334,71 @@ TEST(Validator, importComponents)
...
+    libcellml::ImportSourcePtr imp8 = std::make_shared<libcellml::ImportSource>();
+    imp8->setUrl("yet-another-other-model.xml"); // source used before
+    libcellml::ComponentPtr importedComponent8 = std::make_shared<libcellml::Component>();
+    importedComponent8->setName("a_good_imported_component");
+    importedComponent8->setSourceComponent(imp8, "component_in_that_model");
+    m->addComponent(importedComponent8);
+    v.validateModel(m);
+    EXPECT_EQ(size_t(5), v.errorCount());
+
+    // Invalid: component_ref is not valid html
```
Comment (nickerso) [inline tests/validator/validator.cpp]: its actually the xlink:href that is not a valid URI rather than the component_ref...

**Review by nickerso — APPROVED** [review]
I think the tests are showing the behaviour I'd expect, but some of the comments are a bit confusing. Otherwise happy with these changes.

**Review by agarny — APPROVED** [review]
The code now looks good to me (pending @kerimoyle's replies to @nickerso's questions).

File: tests/validator/validator.cpp (hunk 1 above)
Comment (kerimoyle) [reply tests/validator/validator.cpp]: Sorry - comments were leftovers from previous tests.  You're right, it's permitted to import from the same place.

File: tests/validator/validator.cpp (hunk 2 above)
Comment (kerimoyle) [reply tests/validator/validator.cpp:367]: Added now.

File: tests/validator/validator.cpp (hunk 3 above)
Comment (kerimoyle) [reply tests/validator/validator.cpp]: Yes, it's valid.  Same issue as above where I forgot to update comments when fixing the validator.  Done now :)

File: tests/validator/validator.cpp (hunk 4 above)
Comment (kerimoyle) [reply tests/validator/validator.cpp]: Updated comment

**Comment by kerimoyle** [comment]
@nickerso Have updated comments, you were right :)

**Pull request merged**